### ISSUE TITLE
Catch  and log exception in fromatter_dict, and prompt formatting.

### ISSUE THREAD

**Issue by Carreau** [xonsh/xonsh#1564]
In particular the example in tutorial.rst fais as `$FORMATTER_DICT['test'] = lambda: random.randint(1,9)` does not return a string, and other failures might be hard to debug.

**Comment by adqm** [comment]
Seems to work fine in the `readline` shell, so adding the `prompt-toolkit` label.

**Comment by scopatz** [comment]
This seems to fail for me on both shells:

```console
scopatz@artemis ~ $ import random
scopatz@artemis ~ $ $FORMATTER_DICT['test'] = lambda: random.randint(1,9)
scopatz@artemis ~ $ $PROMPT = '{test} ' + $PROMPT
{test} {env_name:{} }{user}@{hostname} {cwd}{branch_color}{curr_branch: {}} {prompt_end} 
```

**Comment by scopatz** [comment]
Thanks for reporting @Carreau!  I distinctly recall having string conversion originally, so this must have been lost at some point.

**Comment by adqm** [comment]
oh, shoot; i was testing the wrong thing.  my bad.  agree this affects both shells.

**Comment by laerus** [comment]
Testing with ptk atm, it seems that we're calling `prompt.base.partial_format_prompt` 3 times each time we print the prompt. Once in `ptk.shell.PromptToolkitShell.prompt_tokens`, once in `ptk.shell.PromptToolkitShell.rpromt_tokens` and once in `base_shell.BaseShell.settitle`, that seems a little excessive, will try and refactor this out, putting this here as a note.

**Comment by scopatz** [comment]
@laerus - I agree that we could certainly try to do better caching, but somethings like CWD and the last return code actually can and do change each time the prompt is printed.  I think we'd have to clear the cache pretty frequently, but we might be able to cut down on the prompt and rprompt overhead a bit.

**Comment by scopatz** [comment]
Of course this would only help if prompt and rprompt display the same information.  However, this may be more useful if / when we also add a bottom toolbar.  Also, maybe different entries should have a way of flagging how often they need to be revaluated. For example, once (for things like user and host name), every prompt (for repo status, cwd, etc), and every keystroke (for run time validation of input).  This might be a bit of an overhaul.  But it would give us the chance to change the name of `$FORMATTER_DICT` :)

**Comment by laerus** [comment]
Was thinking that `$FORMATTER_DICT` is not a nice name too. Think we should bring events on this?

**Comment by scopatz** [comment]
I don't think this is really event related, personally.  It is a different kind of system.

**Comment by scopatz** [comment]
Here is a failing example:

```
scopatz@artemis ~ $ def bad():
    raise ValueError("I'm bad")
scopatz@artemis ~ $ $PROMPT = '{bad} ' + $PROMPT
{bad} scopatz@artemis ~ $ 
```

**Comment by Ad115** [comment]
> Here is a failing example:
> 
> ```
> scopatz@artemis ~ $ def bad():
>     raise ValueError("I'm bad")
> scopatz@artemis ~ $ $PROMPT = '{bad} ' + $PROMPT
> {bad} scopatz@artemis ~ $ 
> ```

Remember to modify the $PROMPT_FIELDS variables. The complete example then read like this:
```
 def bad(): raise ValueError("I'm bad")
 $PROMPT_FIELDS['bad'] = bad
 $PROMPT = '{bad} ' + $PROMPT
```

**Comment by scopatz** [comment]
So this was closed, but emphasis was added in #3112
